### ISSUE TITLE
SLR does not work for MasterNode

### ISSUE THREAD

**Issue by DavidBoike** [Particular/NServiceBus.Distributor.Msmq#42]
## Who's affected
Any customer attempting to scale out MSMQ as a MasterNode with attached Worker.

## Symptoms
Second level retries does not work when endpoint is deployed as a master node with attached worker and configured with a `MasterNodeConfig` section pointing to the local server. The message is sent to the `EndpointName.Retries` queue, but that queue either does not exist, or exists but the endpoint does not receive messages from it. As a result, faulting messages are passed to SLR but then never reprocessed again, or an exception is thrown that the retries queue does not exist.

## Repro

Bug found by @dvdstelt when working with a customer, and verified by me. Here is a minimal repro:

* Reference NSB 5.2.14, NSB.Host 6.0.0, NSB.Distributor.MSMQ 5.0.4
* Use Code/App.config below.
  * In App.config, set MasterNodeConfig's Node attribute to your own machine name.
* Build, then run from command line `NServiceBus.Host.exe NServiceBus.MSMQMaster`
* Message will fail through FLR and log will say `WARN  NServiceBus.Faults.Forwarder.FaultManager Message with 'c449a8c4-5bc6-4569-a23b-a64a00ad70fa' id has failed FLR and will be handed over to SLR for retry attempt 1.`
  * At this point, the endpoint will not do anything else
  * You'll find the message in the Retries queue, but there's no satellite receiving from that queue.
* If you remove the `MasterNodeConfig` section from the App.config, it will work, advancing through all SLR levels and eventually forwarding the message to the error queue.
  * In this configuration, if you kill the endpoint after the message is handed off to SLR, you'll probably see that the Retries queue is empty. This is because a satellite is receiving messages on the Retries queue faster than you can kill the process, and storing it in timeout storage, which in this case is in-memory. So in this case, there **is** something receiving from the Retries queue.

#### Code

```
namespace SlrBugRepro
{
    public class EndpointConfig : IConfigureThisEndpoint
    {
        public void Customize(BusConfiguration configuration)
        {
            configuration.UsePersistence<InMemoryPersistence>();
            configuration.EnableInstallers(); // So queues are created when run from cmdline
        }
    }

    public class Startup : IWantToRunWhenBusStartsAndStops
    {
        public IBus Bus { get; set; }

        public void Start()
        {
            Bus.Send("SlrBugRepro", new TestCmd());
        }

        public void Stop() { }
    }

    public class TestCmd : ICommand { }

    public class TestHandler : IHandleMessages<TestCmd> {
        public void Handle(TestCmd message)
        {
            Console.WriteLine("Attempting TestCmd");
            throw new Exception("Boom");
        }
    }
}
```

#### App.config

```
<?xml version="1.0" encoding="utf-8" standalone="yes"?>
<configuration>
  <configSections>
    <section name="MessageForwardingInCaseOfFaultConfig" type="NServiceBus.Config.MessageForwardingInCaseOfFaultConfig, NServiceBus.Core" />
    <section name="TransportConfig" type="NServiceBus.Config.TransportConfig, NServiceBus.Core" />
    <section name="SecondLevelRetriesConfig" type="NServiceBus.Config.SecondLevelRetriesConfig, NServiceBus.Core" />
    <section name="MasterNodeConfig" type="NServiceBus.Config.MasterNodeConfig, NServiceBus.Core" />
  </configSections>
  <MessageForwardingInCaseOfFaultConfig ErrorQueue="error" />
  <TransportConfig MaxRetries="2" MaximumConcurrencyLevel="1" MaximumMessageThroughputPerSecond="0" />
  <SecondLevelRetriesConfig Enabled="true" NumberOfRetries="3" TimeIncrease="00:00:5" />
  <MasterNodeConfig Node="DAVIDBOIKE0C30" />
</configuration>
```

**Comment by omega1970** [comment]
We have located the problem !!!

The problem is in the configuration file, when running a combined distributor / worker endpoint, you are not allowed to use the:
	  <MasterNodeConfig Node="machine-name-of-combined-distributor" />

This setting changes the behavior of the distributor on the combined endpoint, causing the queues ‘retries’ and ‘worker’ not being created and processed. So in short, this setting should be ignored when present or clearly documented that this setting must not be used on a combined endpoint !!

**Comment by DavidBoike** [comment]
@omega1970 That's a valid workaround for now, but in my view you should be able to do that. One of the selling points of the distributor is that you should be able to take exactly the same binary+config package, deploy it to multiple servers, and control whether it's a master/worker/distributor using only the profile name when you install the endpoint. You shouldn't have to do any special DevOps magic to change the config for the master node.

@danielmarbach and I looked through the code just now and think we may have found a solution. As soon as his fix is built on TeamCity I'll try it out on my local repro.

**Comment by DavidBoike** [comment]
Fix is merged to master and will be 5.0.5 - need to find time tomorrow to release.

**Comment by DavidBoike** [comment]
Released: https://github.com/Particular/NServiceBus.Distributor.Msmq/releases/tag/5.0.5
